lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,260 @@
+        _validate_connection_params(conn_kwargs)
+
+        return snowflake.connector.connect(**conn_kwargs)
+
+    def query(
+        self,
+        sql: str,
+        *,  # keyword-only arguments:
+        ttl: float | int | timedelta | None = None,
+        show_spinner: bool | str = "Running `snowflake.query(...)`.",
+        params=None,
+    ) -> pd.DataFrame:
+        """Run a read-only SQL query.
+
+        This method implements both query result caching (with caching behavior
+        identical to that of using ``@st.cache_data``) as well as simple error handling/retries.
+
+        .. note::
+            Queries that are run without a specified ttl are cached indefinitely.
+
+        Parameters
+        ----------
+        sql : str
+            The read-only SQL query to execute.
+        ttl : float, int, timedelta or None
+            The maximum number of seconds to keep results in the cache, or
+            None if cached results should not expire. The default is None.
+        show_spinner : boolean or string
+            Enable the spinner. The default is to show a spinner when there is a
+            "cache miss" and the cached resource is being created. If a string, the value
+            of the show_spinner param will be used for the spinner text.
+        params : list, tuple, dict or None
+            List of parameters to pass to the execute method. The syntax used to pass
+            parameters is database driver dependent. Check your database driver
+            documentation for which of the five syntax styles, described in `PEP 249
+            paramstyle <https://peps.python.org/pep-0249/#paramstyle>`_, is supported.
+            Default is None.
+
+        Returns
+        -------
+        pd.DataFrame
+            The result of running the query, formatted as a pandas DataFrame.
+
+        Example
+        -------
+        >>> import streamlit as st
+        >>>
+        >>> conn = st.connection("snowflake")
+        >>> df = conn.query("select * from pet_owners")
+        >>> st.dataframe(df)
+        """
+        # TODO(vdonato): Make our error handling more specific if possible. This may be
+        # difficult to do given the limited documentation on the different connector
+        # error subclasses + how many there are.
+        from snowflake.connector import Error as SnowflakeError
+        from tenacity import (
+            retry,
+            retry_if_exception_type,
+            stop_after_attempt,
+            wait_fixed,
+        )
+
+        @retry(
+            after=lambda _: self.reset(),
+            stop=stop_after_attempt(3),
+            reraise=True,
+            retry=retry_if_exception_type(SnowflakeError),
+            wait=wait_fixed(1),
+        )
+        @cache_data(
+            show_spinner=show_spinner,
+            ttl=ttl,
+        )
+        def _query(sql: str) -> pd.DataFrame:
+            cur = self._instance.cursor()
+            cur.execute(sql, **(params or {}))
```

Comment:
I think this is incorrect - params value should be passed as `parameters` arg, not expanded as kwargs. I get an error when I try to use it:

```python
df = conn.query("SELECT %(col1)s, %(col2)s",
                ttl=0,
                params={'col1': 789, 'col2': 'test string3'})
```

`TypeError: SnowflakeCursor.execute() got an unexpected keyword argument 'col1'`

https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-api#id2

Reply:
(and also get an error when passing a tuple and using qmark syntax)

Reply:
Whoops, I misunderstood what the notation in the docs meant -- will fix